File: tools/excel/src/test/java/top/bettercode/summer/tools/excel/ExcelTest.java
package top.bettercode.summer.tools.excel;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import top.bettercode.summer.tools.lang.util.ArrayUtil;
import top.bettercode.summer.tools.lang.util.StringUtil;

/**
 * @author Peter Wu
 * @since 0.0.1
 */
public class ExcelTest {

  @BeforeEach
  public void setUp() {

  }

  private final ExcelField<DataBean, ?>[] excelFields = ArrayUtil.of(
      ExcelField.index("序号"),
      ExcelField.of("编码1", DataBean::getIntCode),
      ExcelField.of("编码2", DataBean::getInteger),
      ExcelField.of("编码3", DataBean::getLongl),
      ExcelField.of("编码4", DataBean::getDoublel),
      ExcelField.of("编码5", DataBean::getFloatl),
      ExcelField.of("编码6", DataBean::getName),
      ExcelField.of("编码7", DataBean::getDate),
      ExcelField.of("编码8", DataBean::getNum)
  );

  @Test
  public void testExport() throws IOException {

    List<DataBean> list = new ArrayList<>();
    for (int i = 0; i < 8; i++) {
      DataBean bean = new DataBean(i);
      list.add(bean);
    }
    long s = System.currentTimeMillis();
    ExcelExport.of("build/export.xlsx").sheet("表格")
        .setData(list, excelFields).finish();
    long e = System.currentTimeMillis();
    System.err.println(e - s);
    openExcel("build/export.xlsx");
  }

  private static void openExcel(String x) throws IOException {
    Runtime.getRuntime().exec(new String[]{"xdg-open", System.getProperty("user.dir") + "/" + x});
  }

  private final ExcelField<DataBean, ?>[] excelMergeFields = ArrayUtil.of(
      ExcelField.<DataBean, Integer>index("序号").mergeBy(DataBean::getIntCode),
      ExcelField.of("编码", DataBean::getIntCode).mergeBy(DataBean::getIntCode),
      ExcelField.of("编码B", DataBean::getInteger).mergeBy(DataBean::getInteger),
      ExcelField.of("名称", from -> new String[]{"abc", "1"}),
      ExcelField.of("描述", DataBean::getName),
      ExcelField.of("描述C", DataBean::getDate)
  );

  @Test
  public void testMergeExport() throws IOException {
    List<DataBean> list = new ArrayList<>();
    for (int i = 0; i < 22; i++) {
      DataBean bean = new DataBean(i);
      list.add(bean);
    }
    long s = System.currentTimeMillis();
    ExcelExport.of("build/export.xlsx").sheet("表格")
        .setMergeData(list, excelMergeFields).finish();
    long e = System.currentTimeMillis();
    System.err.println(e - s);
    openExcel("build/export.xlsx");
  }


  @Test
  public void testImport() throws Exception {
//    testExport();
    List<DataBean> list = ExcelImport.of("build/template.xlsx").setColumn(1)
        .getData(excelFields);
    System.out.println(StringUtil.valueOf(list, true));
    System.err.println(list.size());
  }

  @Test
  public void testTemplate() throws IOException {
    ExcelExport.of("build/template.xlsx").sheet("表格1").dataValidation(1, "1,2,3")
        .template(excelFields)
        .finish();
    openExcel("build/template.xlsx");
  }
}

File: tools/excel/src/test/kotlin/top/bettercode/summer/tools/excel/ExcelTest.kt
package top.bettercode.summer.tools.excel

import org.junit.jupiter.api.*
import org.springframework.core.io.ClassPathResource
import top.bettercode.summer.tools.excel.read.CellGetter
import top.bettercode.summer.tools.excel.read.ExcelReader
import top.bettercode.summer.tools.excel.read.RowGetter
import top.bettercode.summer.tools.excel.write.CellSetter
import top.bettercode.summer.tools.excel.write.ExcelWriter
import top.bettercode.summer.tools.excel.write.RowSetter
import top.bettercode.summer.tools.lang.util.StringUtil.json

/**
 * @author Peter Wu
 * @since 0.0.1
 */
@TestMethodOrder(MethodOrderer.OrderAnnotation::class)
open class ExcelTest {

    protected open var nameSuff: String = "1"

    private val rowSetter = RowSetter.of(
        CellSetter.index<DataBean, Any?>("序号"),
        CellSetter.of("编码1", DataBean::intCode),
        CellSetter.of("编码2", DataBean::integer).comment("批注"),
        CellSetter.of("编码3", DataBean::longl),
        CellSetter.of("编码4", DataBean::doublel).comment("批注2"),
        CellSetter.of("编码5", DataBean::floatl),
        CellSetter.of("编码6", DataBean::name),
        CellSetter.of("编码7") { obj: DataBean -> obj.date },
        CellSetter.of("编码8") { obj: DataBean -> obj.num }
    )

    @Test
    open fun testExport() {
        val list: MutableList<DataBean> = ArrayList()
        for (i in 0..7) {
            val bean = DataBean(i)
            list.add(bean)
        }
        val s = System.currentTimeMillis()
        val filename = "build/testExport$nameSuff.xlsx"
        excelExport(filename).use {
            it.sheet("表格")
                .setData(list, rowSetter)
        }
        val e = System.currentTimeMillis()
        System.err.println(e - s)
        ExcelTestUtil.openExcel(filename)
    }

    private val rowSetter2 = RowSetter.of(
        CellSetter.index<DataBean, Int?>("序号").mergeBy { obj: DataBean -> obj.intCode },
        CellSetter.of("编码") { obj: DataBean -> obj.intCode }
            .mergeBy { obj: DataBean -> obj.intCode },
        CellSetter.of("编码B") { obj: DataBean -> obj.integer }
            .mergeBy { obj: DataBean -> obj.integer },
        CellSetter.of("名称") { arrayOf("abc", "1") },
        CellSetter.of("描述") { obj: DataBean -> obj.name },
        CellSetter.of("描述C") { obj: DataBean -> obj.date }
    )

    @Test
    open fun testMergeExport() {
        val list: MutableList<DataBean> = ArrayList()
        for (i in 0..21) {
            val bean = DataBean(i)
            list.add(bean)
        }
        val s = System.currentTimeMillis()
        val filename = "build/testMergeExport$nameSuff.xlsx"
        excelExport(filename).use {
            it.sheet("表格")
                .setData(list, rowSetter2)
        }
        val e = System.currentTimeMillis()
        System.err.println(e - s)
        ExcelTestUtil.openExcel(filename)
    }

    private val rowGetter  = RowGetter.of(
        CellGetter.of("编码1", DataBean::intCode),
        CellGetter.of("编码2", DataBean::integer),
        CellGetter.of("编码3", DataBean::longl),
        CellGetter.of("编码4", DataBean::doublel),
        CellGetter.of("编码5", DataBean::floatl),
        CellGetter.of("编码6", DataBean::name)
    )

    @Order(1)
    @Test
    fun testImport() {
        ExcelReader.of(ClassPathResource("template.xlsx").inputStream).use {
            it.column(1)
            val list = it.getData<DataBean, DataBean>(rowGetter)
            println(json(list, true))
            System.err.println(list.size)
            Assertions.assertEquals(3, list.size)
        }
    }

    @Order(0)
    @Test
    open fun testTemplate() {
        val filename = "build/template$nameSuff.xlsx"
        excelExport(filename).use {
            it.sheet("表格1").dataValidation(1, "1", "2", "3")
                .template(rowSetter)
        }
        ExcelTestUtil.openExcel(filename)
    }

    protected open fun excelExport(filename: String) = ExcelWriter.of(filename)

}